main/src/ca/uwaterloo/flix/language/phase/jvm/GenLazyClasses.scala

```diff
@@ -0,0 +1,274 @@
+  }
+
+  /**
+   * !OBS!: This code was copied from AsmOps to omit the last cast in the function.
+   *
+   * Emits code to call a closure (not in tail position). fType is the type of the called
+   * closure. argsType is the type of its arguments, and resultType is the type of its result.
+   */
+  private def compileClosureApplication(visitor: MethodVisitor, valueType: MonoType)(implicit root: Root, flix: Flix) = {
+    val fType = MonoType.Arrow(List(MonoType.Unit), valueType)
+    // Type of the continuation interface
+    val cont = JvmOps.getContinuationInterfaceType(fType)
+    // Label for the loop
+    val loop = new Label
+    visitor.visitFieldInsn(PUTFIELD, JvmName.Context.toInternalName, "continuation", JvmType.Object.toDescriptor)
+    // Begin of the loop
+    visitor.visitLabel(loop)
+    // Getting `continuation` field on `Context`
+    visitor.visitVarInsn(ALOAD, 1)
+    visitor.visitFieldInsn(GETFIELD, JvmName.Context.toInternalName, "continuation", JvmType.Object.toDescriptor)
+    // Setting `continuation` field of global to `null`
+    visitor.visitVarInsn(ALOAD, 1)
+    visitor.visitInsn(ACONST_NULL)
+    visitor.visitFieldInsn(PUTFIELD, JvmName.Context.toInternalName, "continuation", JvmType.Object.toDescriptor)
+    // Cast to the continuation
+    visitor.visitTypeInsn(CHECKCAST, cont.name.toInternalName)
+    // Duplicate
+    visitor.visitInsn(DUP)
+    // Save it on the IFO local variable
+    visitor.visitVarInsn(ASTORE, 2)
+    // Call invoke
+    visitor.visitVarInsn(ALOAD, 1)
+    visitor.visitMethodInsn(INVOKEINTERFACE, cont.name.toInternalName, "invoke", AsmOps.getMethodDescriptor(List(JvmType.Context), JvmType.Void), true)
+    // Getting `continuation` field on `Context`
+    visitor.visitVarInsn(ALOAD, 1)
+    visitor.visitFieldInsn(GETFIELD, JvmName.Context.toInternalName, "continuation", JvmType.Object.toDescriptor)
+    visitor.visitJumpInsn(IFNONNULL, loop)
+    // Load IFO from local variable and invoke `getResult` on it
+    visitor.visitVarInsn(ALOAD, 2)
+    visitor.visitMethodInsn(INVOKEINTERFACE, cont.name.toInternalName, "getResult", AsmOps.getMethodDescriptor(Nil, JvmOps.getErasedJvmType(valueType)), true)
+  }
+
+  /**
+   * The force method takes a context as argument to call the expression closure in.
+   * The result of the expression given in the constructor is then returned.
+   * This is only actually evaluated the first time, and saved to return directly
+   * afterwards.
+   *
+   * If lazy has associated type of Obj, the returned object needs to be casted
+   * to whatever expected type.
+   */
+  private def compileForceMethod(visitor: ClassWriter, classType: JvmType.Reference, valueType: MonoType)(implicit root: Root, flix: Flix): Unit = {
+    /*
+    force(context) :=
+
+    lock(this)
+    if (!this.initialized) {
+      this.value = this.expression()
+      this.initialized = true
+    }
+    tpe result = this.value
+    unlock(this)
+    return result
+     */
```

Comment:
I would be nice if our new backend could have a DSL that end up looking like the above!